File: testopengl/src/testopengl/OpenGLHelper.cpp
#include "OpenGLHelper.h"
#include <assert.h>
#include <stdio.h>


namespace OpenGLHelper {
void test() {

}

void outputCompileShader(GLuint shader)
{
	GLint status = 0;
	glGetShaderiv(shader, GL_COMPILE_STATUS, &status);
	if (GL_FALSE == status)
	{
		const int MAX_LOG_LENGTH = 512;
		GLint len = 0;
		glGetShaderiv(shader, GL_INFO_LOG_LENGTH, &len);
		if (len < MAX_LOG_LENGTH)
		{
			GLchar szBuf[MAX_LOG_LENGTH] = { 0 };
			int len = MAX_LOG_LENGTH;
			glGetShaderInfoLog(shader, MAX_LOG_LENGTH, nullptr, szBuf);
			printf("%s\n", szBuf);
		}
		else
		{
			GLchar *buf = new GLchar[len + 1];
			glGetShaderInfoLog(shader, len, nullptr, buf);
			buf[len] = '\0';
			printf("%s\n", buf);
			delete[]buf;
		}
		assert(GL_FALSE != status);
	}
}

void outputProgramLog(GLuint program)
{
	GLint status = 0;
	glGetProgramiv(program, GL_LINK_STATUS, &status);
	if (GL_FALSE == status)
	{
		const int MAX_LOG_LENGTH = 512;
		GLint len = 0;
		glGetProgramiv(program, GL_INFO_LOG_LENGTH, &len);
		if (len < MAX_LOG_LENGTH)
		{
			GLchar szBuf[MAX_LOG_LENGTH] = { 0 };
			glGetProgramInfoLog(program, MAX_LOG_LENGTH, nullptr, szBuf);
			printf("%s\n", szBuf);
		}
		else
		{
			GLchar *buf = new GLchar[len + 1];
			glGetProgramInfoLog(program, len, nullptr, buf);
			buf[len] = '\0';
			printf("%s\n", buf);
			delete[]buf;
		}
		assert(false);
	}
}

GLenum attachShader(GLuint program, GLenum type, const char *source, GLint len) {
	GLenum ret = GL_NO_ERROR;
	GLuint shader = 0;

	do
	{
		if (0 == (shader = glCreateShader(type))) {
			ret = glGetError();
			break;
		}
		GLint lens[] = { len };
		glShaderSource(shader, 1, &source, (len > 0 ? lens : nullptr));
		if (GL_NO_ERROR != (ret = glGetError())) {
			break;
		}
		glCompileShader(shader);
		outputCompileShader(shader);
		if (GL_NO_ERROR != (ret = glGetError())) {
			break;
		}
		glAttachShader(program, shader);
		if (GL_NO_ERROR != (ret = glGetError())) {
			break;
		}
	} while (false);

	if (0 != shader) {
		glDeleteShader(shader);
		shader = 0;
	}
	return ret;
}
};

File: android/testcamera/app/src/main/java/com/test/testcamera/OpenGLHelper.java
package com.test.testcamera;

import android.opengl.GLES30;
import android.util.Log;

public class OpenGLHelper {
    public static void attachShader(int program, int shaderType, String shaderSource) throws Exception{
        int shader = 0;
        try {
            shader = GLES30.glCreateShader(shaderType);
            checkError();
            GLES30.glShaderSource(shader, shaderSource);
            checkError();
            GLES30.glCompileShader(shader);
            checkError();
            outputCompileShader(shader);
            GLES30.glAttachShader(program, shader);
            checkError();
        }finally {
            if (0 != shader){
                GLES30.glDeleteShader(shader);
            }
        }
    }

    public static void checkError() throws MyException{
        int err = GLES30.glGetError();
        if (GLES30.GL_NO_ERROR != err){
            throw new MyException(err);
        }
    }

    public static class MyException extends Exception{
        private int mCode;
        public MyException(int code){
            super();
            mCode = code;
        }

        int getCode(){
            return mCode;
        }
    }

    private static int[] status_params = new int[1];
    public static void outputCompileShader(int shader){
        GLES30.glGetShaderiv(shader, GLES30.GL_COMPILE_STATUS, status_params, 0);
        if (GLES30.GL_TRUE != status_params[0]){
            String log = GLES30.glGetShaderInfoLog(shader);
            Log.e("GLSL30", "shader:" + log);
        }
    }

    public static void outputProgramLog(int program){
        GLES30.glGetProgramiv(program, GLES30.GL_LINK_STATUS, status_params, 0);
        if (GLES30.GL_TRUE != status_params[0]){
            String log = GLES30.glGetProgramInfoLog(program);
            Log.e("GLSL30", "Program:" + log);
        }
    }
}
